### PR TITLE
msg/async/rdma: refactor tx handle flow to get rid of locks

### PR DESCRIPTION
**Pull request by yuyuyu101**
previously Dispatcher thread will poll both rx and tx events, then dispatch these events to RDMAWorker and RDMAConnectedSocketImpl.

Actually tx event handling is a lightweight task and we make these handling inline now. rx event dispatching is still working.

Another change is adding tx cq to make event polling separated.

removing lots of codes yet.

### CODE REVIEW

**Comment by yuyuyu101** [comment]
@Adirl @orendu now the whole io flow will make much clear... dispatcher thread handle tx cq directly, dispatch rx cq to workers. less locks, less hand on

**Comment by yuyuyu101** [comment]
still need to solve failed test case

File: src/msg/async/rdma/RDMAStack.cc [hunk 1]
```diff
@@ -472,16 +471,13 @@ void RDMAWorker::post_tx_buffer(std::vector<Chunk*> &chunks)
...
   std::set<RDMAConnectedSocketImpl*> done;
-  Mutex::Locker l(lock);
   while (!pending_sent_conns.empty()) {
     RDMAConnectedSocketImpl *o = pending_sent_conns.front();
     pending_sent_conns.pop_front();
     if (!done.count(o)) {
-      lock.Unlock();
       done.insert(o);
       ssize_t r = o->submit(false);
       ldout(cct, 20) << __func__ << " sent pending bl socket=" << o << " r=" << r << dendl;
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: looks like false param to submit has no effect 
?!

File: src/msg/async/rdma/RDMAStack.cc [hunk 2]
```diff
@@ -125,16 +133,68 @@ void RDMADispatcher::polling()
...
-  int ret = 0;
+  int r = 0;
 
   while (true) {
-    int n = rx_cq->poll_cq(MAX_COMPLETIONS, wc);
-    if (!n) {
+    int tx_ret = tx_cq->poll_cq(MAX_COMPLETIONS, wc);
+    if (tx_ret > 0) {
+      ldout(cct, 20) << __func__ << " pool tx completion queue got " << tx_ret
+                     << " responses."<< dendl;
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: i guess you mean "poll "

File: src/msg/async/rdma/RDMAStack.cc [hunk 3]
```diff
@@ -125,16 +133,68 @@ void RDMADispatcher::polling()
...
+      handle_tx_event(wc, tx_ret);
+    }
+
+    int rx_ret = rx_cq->poll_cq(MAX_COMPLETIONS, wc);
+    if (rx_ret > 0) {
+      ldout(cct, 20) << __func__ << " pool rt completion queue got " << rx_ret
+                     << " responses."<< dendl;
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: i guess you mean "poll rx"

File: src/msg/async/rdma/RDMAStack.cc [hunk 4]
```diff
@@ -125,16 +133,68 @@ void RDMADispatcher::polling()
...
+  int r = 0;
 
   while (true) {
-    int n = rx_cq->poll_cq(MAX_COMPLETIONS, wc);
-    if (!n) {
+    int tx_ret = tx_cq->poll_cq(MAX_COMPLETIONS, wc);
+    if (tx_ret > 0) {
+      ldout(cct, 20) << __func__ << " pool tx completion queue got " << tx_ret
+                     << " responses."<< dendl;
+      handle_tx_event(wc, tx_ret);
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: handle here? on same thread?

Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: you are correct, it's ok to handle Tx here, 
i see you removed some code from post_tx_buffer() that i was worried about

File: src/msg/async/rdma/RDMAStack.cc [hunk 5]
```diff
@@ -335,6 +316,75 @@ void RDMADispatcher::handle_post_fork()
...
+      if (conn && conn->is_connected()) {
+        ldout(cct, 25) << __func__ << " qp state is : " << conn->get_qp_state() << dendl;//wangzhi
+        conn->fault();
+      } else {
+        ldout(cct, 1) << __func__ << " missing qp_num=" << response->qp_num << " discard event" << dendl;
+      }
+    }
+
+    // FIXME: why not tx?
+    if (ib->get_memory_manager()->is_tx_buffer(chunk->buffer))
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: maybe assert it's Tx cqe like you did for Rx
assert(wc[i].opcode == IBV_WC_RECV);

File: src/msg/async/rdma/RDMAStack.cc [hunk 6]
```diff
@@ -335,6 +316,75 @@ void RDMADispatcher::handle_post_fork()
...
+void RDMADispatcher::post_tx_buffer(std::vector<Chunk*> &chunks)
+{
+  if (chunks.empty())
+    return ;
+
+  inflight -= chunks.size();
+  ib->get_memory_manager()->return_tx(chunks);
+  ldout(cct, 30) << __func__ << " release " << chunks.size()
+                 << " chunks, inflight " << inflight << dendl;
+  notify_pending_workers();
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: why is that needed?

File: src/msg/async/rdma/RDMAStack.cc [hunk 7]
```diff
@@ -437,39 +454,25 @@ int RDMAWorker::reserve_message_buffer(RDMAConnectedSocketImpl *o, std::vector<C
 {
   assert(center.in_thread());
   int r = infiniband->get_tx_buffers(c, bytes);
-  if (r > 0) {
-    stack->get_dispatcher()->inflight += r;
-    ldout(cct, 30) << __func__ << " reserve " << r << " chunks, inflight " << dispatcher->inflight << dendl;
+  assert(r >= 0);
+  size_t got = infiniband->get_memory_manager()->get_tx_chunk_size() * r;
+  ldout(cct, 30) << __func__ << " need " << bytes << " bytes, reserve " << got << " bytes, inflight " << dispatcher->inflight << dendl;
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: maybe 
ldout(cct, 30) << __func__ << " need " << bytes << " bytes, got " << got << "registered bytes, inflight " << dispatcher->inflight << dendl;

**Comment by Adirl** [comment]
attaching osd log of this code running on real cluster 
FYI 

[ceph-osd.0 (2).txt](https://github.com/ceph/ceph/files/806843/ceph-osd.0.2.txt)

**Comment by Adirl** [comment]
@yuyuyu101 
found few of these in osd log:
2017-02-28 12:57:50.671156 7f2b1c928700  1 RDMAStack handle_tx_event not tx buffer, chunk 0x7f2b2fb74000

shouldn't happen, right ?

File: src/msg/async/rdma/RDMAStack.cc [hunk 8]
```diff
@@ -434,114 +452,49 @@ int RDMAWorker::connect(const entity_addr_t &addr, const SocketOptions &opts, Co
 
 int RDMAWorker::reserve_message_buffer(RDMAConnectedSocketImpl *o, std::vector<Chunk*> &c, size_t bytes)
```
Comment (Adirl) [inline src/msg/async/rdma/RDMAStack.cc]: can we call it: "get_reged_mem()"

**Comment by Adirl** [comment]
@yuyuyu101 
is this last commit : msg/async: must destruct ConnectedSocket in the his own thread  …
solving the issue i saw ?  (RDMAStack handle_tx_event not tx buffer)

File: src/msg/async/rdma/RDMAStack.cc (hunk 6 above)
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAStack.cc]: wake up worker to handle pending message which is eager for tx buffer

**Comment by yuyuyu101** [comment]
@Adirl fixed now. plz check again.

**Comment by Adirl** [comment]
@yuyuyu101 
looks ok so far

[root@r-dcs69 ~]# rados bench -p rbd 10 write --no-cleanup                                                                                         
  sec Cur ops   started  finished  avg MB/s  cur MB/s last lat(s)  avg lat(s)                                 
    0       0         0         0         0         0           -           0                                 
    1      16        32        16   63.9953        64     0.88772    0.500009    

[root@r-dcs69 ~]#  rados bench -p rbd 10 rand
  sec Cur ops   started  finished  avg MB/s  cur MB/s last lat(s)  avg lat(s)
    0       0         0         0         0         0           -           0
    1      16       773       757   3027.35      3028  0.00602852   0.0198304

**Comment by Adirl** [comment]
i see some memory missing at worker:
    "AsyncMessenger::RDMAWorker-1": {
        "tx_no_mem": 8,
        "tx_parital_mem": 7,
        "tx_failed_post": 0,

**Comment by Adirl** [comment]
./fio --ioengine=rbd --rw=rw --bs=1M --numjobs=1 --clientname=admin --pool=rbd --iodepth=1 --rbdname=fio_test --name=1 --time_based --runtime=20 --verify=crc32c --do_verify=1

  READ: bw=15.8MiB/s (16.5MB/s), 15.8MiB/s-15.8MiB/s (16.5MB/s-16.5MB/s), io=314MiB (329MB), run=20004-20004msec
  WRITE: bw=16.6MiB/s (17.3MB/s), 16.6MiB/s-16.6MiB/s (17.3MB/s-17.3MB/s), io=330MiB (346MB), run=20004-20004msec


also counters are clean 
hooray!
